src/dhcpmon/src/dhcp_device.c

```diff
@@ -0,0 +1,435 @@
+                                              dhcp_device_counters_t *counters_snapshot)
+{
+    dhcp_mon_status_t rv = DHCP_MON_STATUS_HEALTHY;
+
+    if ((counters[DHCP_RX].discovery == counters_snapshot[DHCP_RX].discovery) &&
+        (counters[DHCP_RX].offer == counters_snapshot[DHCP_RX].offer) &&
+        (counters[DHCP_RX].request == counters_snapshot[DHCP_RX].request) &&
+        (counters[DHCP_RX].ack == counters_snapshot[DHCP_RX].ack)) {
+        rv = DHCP_MON_STATUS_INDETERMINATE;
+    } else {
+        // if we have rx DORA then we should have corresponding tx DORA (DORA being relayed)
+        if (((counters[DHCP_RX].discovery > counters_snapshot[DHCP_RX].discovery ) &&
+             (counters[DHCP_TX].discovery <= counters_snapshot[DHCP_TX].discovery)    ) ||
+            ((counters[DHCP_RX].offer > counters_snapshot[DHCP_RX].offer ) &&
+             (counters[DHCP_TX].offer <= counters_snapshot[DHCP_TX].offer)            ) ||
+            ((counters[DHCP_RX].request > counters_snapshot[DHCP_RX].request ) &&
+             (counters[DHCP_TX].request <= counters_snapshot[DHCP_TX].request)        ) ||
+            ((counters[DHCP_RX].ack > counters_snapshot[DHCP_RX].ack ) &&
+             (counters[DHCP_TX].ack <= counters_snapshot[DHCP_TX].ack)                )) {
+            rv = DHCP_MON_STATUS_UNHEALTHY;
+        }
+    }
+
+    return rv;
+}
+
+/**
+ * @code dhcp_print_counters(counters);
+ *
+ * @brief prints DHCP counters to sylsog.
+ *
+ * @param counters  interface counter
+ */
+static void dhcp_print_counters(dhcp_device_counters_t *counters)
+{
+    syslog(LOG_NOTICE, "DHCP Discover rx: %lu, tx:%lu, Offer rx: %lu, tx:%lu, Request rx: %lu, tx:%lu, ACK rx: %lu, tx:%lu\n",
+           counters[DHCP_RX].discovery, counters[DHCP_TX].discovery,
+           counters[DHCP_RX].offer, counters[DHCP_TX].offer,
+           counters[DHCP_RX].request, counters[DHCP_TX].request,
+           counters[DHCP_RX].ack, counters[DHCP_TX].ack);
+}
+
+/**
+ * @code init_libpcap(context, intf, snaplen, timeout_ms);
+ *
+ * @brief initializes libpacp for a device (interface)
+ *
+ * @param context           pointer to device (interface) context
+ * @param intf              interface name
+ * @param snaplen           length of packet capture
+ * @aparam timeout_ms       timeout for libpacp to report packets captured
+ *
+ * @return 0 on success, otherwise for failure
+ */
+static int init_libpcap(dhcp_device_context_t *context, const char *intf, int snaplen, int timeout_ms)
+{
+    int rv = -1;
+
+    do {
+        strncpy(context->intf, intf, sizeof(context->intf) - 1);
+        context->intf[sizeof(context->intf) - 1] = '\0';
+
+        context->cap = pcap_open_live(context->intf, snaplen, DHCP_PROMISC_MODE, timeout_ms, context->err_buf);
+        if (context->cap == NULL) {
+            syslog(LOG_ALERT, "pcap_open_live(): %s", context->err_buf);
+            break;
+        }
+
+        if (pcap_compile(context->cap, &context->fp, DHCP_BPF,
```

Comment:
can we compile once, from command line and then use compiled binary here?

Reply:
The pcap_compile accepts pcap_t struct as a first argument. From online reading, it needs the link layer information. I thought about compiling it once during init, however i refrained from doing so as the pcap_t struct is per interface. Also, do not know what the command line equivalent if any.

Reply:
tcpdump -d -dd -ddd does that. Check that the bpf output doesn't rely on interface name.
Usually I don't use libpcap, I compile bpf by tcpdump, then open a socket in RAW mode, bind it to an interface, and bind compiled bpf program to the socket. After that it's possible to use poll() to check all sockets at the same time, without pthreads. Which significantly simplify application.
```
pavelsh@pavelsh-vm:~/disk2/201811/sonic-buildimage/src/lldpd/patch$ sudo tcpdump -dd ether proto 0x8899
{ 0x28, 0, 0, 0x0000000c },
{ 0x15, 0, 1, 0x00008899 },
{ 0x6, 0, 0, 0x00040000 },
{ 0x6, 0, 0, 0x00000000 },
pavelsh@pavelsh-vm:~/disk2/201811/sonic-buildimage/src/lldpd/patch$ sudo tcpdump -ddd ether proto 0x8899
4
40 0 0 12
21 0 1 34969
6 0 0 262144
6 0 0 0
pavelsh@pavelsh-vm:~/disk2/201811/sonic-buildimage/src/lldpd/patch$ sudo tcpdump -d ether proto 0x8899  
(000) ldh      [12]
(001) jeq      #0x8899          jt 2    jf 3
(002) ret      #262144
(003) ret      #0
```

Reply:
I will upload a patch that does eliminate libpcap in favor of raw sockets and libeevent. Let's vote which one to keep.